browser/brave_ads/brave_ads_host.cc

```diff
@@ -0,0 +1,162 @@
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  brave_rewards::RewardsService* rewards_service =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile);
+  if (!rewards_service || !ads_service || !ads_service->IsSupportedLocale()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (ads_service->IsEnabled()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  if (callback_) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  if (!browser) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  callback_ = std::move(callback);
+
+  rewards_service->StartProcess(base::DoNothing());
+  rewards_service_observation_.Observe(rewards_service);
+  action_ui_observation_.Observe(extensions::BraveActionAPI::Get(browser));
+  initiated_browser_ = browser;
+
+  if (!ShowRewardsPopup(browser, profile)) {
+    Reset();
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::WebContentsDestroyed() {
+  Reset();
+
+  if (callback_) {
+    std::move(callback_).Run(false);
+  }
+}
+
+void BraveAdsHost::OnRewardsEnabled() {
+  DCHECK(callback_);
+
+  Reset();
+
+  std::move(callback_).Run(true);
+}
+
+void BraveAdsHost::OnAdsEnabled(brave_rewards::RewardsService* rewards_service,
+                                bool ads_enabled) {
+  DCHECK(rewards_service);
+  DCHECK(callback_);
+
+  Reset();
+
+  std::move(callback_).Run(ads_enabled);
+}
+
+void BraveAdsHost::OnActionUIClosed(Browser* browser,
+                                    const std::string& extension_id) {
+  DCHECK(callback_);
+  if (extension_id != brave_rewards_extension_id) {
+    return;
+  }
+
+  const Browser* current_browser =
+      chrome::FindBrowserWithWebContents(web_contents());
+  if (browser != current_browser && browser != initiated_browser_) {
+    return;
+  }
+
+  Reset();
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  DCHECK(profile);
+
+  const AdsService* ads_service = AdsServiceFactory::GetForProfile(profile);
+  bool ads_enabled = false;
+  if (ads_service) {
+    ads_enabled = ads_service->IsEnabled();
+  }
+  std::move(callback_).Run(ads_enabled);
+}
+
+bool BraveAdsHost::ShowRewardsPopup(Browser* browser, Profile* profile) {
+  auto* extension_service =
+      extensions::ExtensionSystem::Get(profile)->extension_service();
+  if (!extension_service) {
+    return false;
+  }
+
+  static_cast<extensions::BraveComponentLoader*>(
+      extension_service->component_loader())
+      ->AddRewardsExtension();
+
+  std::string error;
+  return extensions::BraveActionAPI::ShowActionUI(
+      browser, brave_rewards_extension_id,
+      std::make_unique<std::string>(kAdsEnableRelativeUrl), &error);
+}
+
+void BraveAdsHost::Reset() {
```

Comment:
curious why don't we reset `callback_`

Reply:
callback_ is called after each Reset() call with different arguments (true or false)

Reply:
yeah I mean maybe we could combine these two calls somehow, b/c the current logic seems error-prone